### DISCUSSION TITLE
RP2040 Support

### DISCUSSION THREAD

**Discussion by jp112sdl** [pa-pa/AskSinPP#294] [Ideas]
Dann fange ich mal die Diskussion hier an...

@pa-pa Ich hänge zurzeit noch an einem letzten Problem beim Kompilieren.

https://github.com/pa-pa/AskSinPP/blob/297bfb84605b666061837190acc5232b7fe6f5ef/AskSinPP.h#L91

endet immer in
```
error: invalid conversion from 'int' to 'const void*' [-fpermissive]
   91 |       *dest = pgm_read_byte(adr + i);
      |                             ~~~~^~~
      |                                 |
      |                                 int
In file included from /Applications/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/rp2040/hardware/rp2040/1.12.0/cores/rp2040/api/String.h:33,
                 from /Applications/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/rp2040/hardware/rp2040/1.12.0/cores/rp2040/api/IPAddress.h:24,
                 from /Applications/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/rp2040/hardware/rp2040/1.12.0/cores/rp2040/api/ArduinoAPI.h:30,
                 from /Applications/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/rp2040/hardware/rp2040/1.12.0/cores/rp2040/Arduino.h:28,
                 from ../sloeber.ino.cpp:7:
/Applications/Sloeber.app/Contents/Eclipse/arduinoPlugin/packages/rp2040/hardware/rp2040/1.12.0/cores/rp2040/api/deprecated-avr-comp/avr/pgmspace.h:116:55: note:   initializing argument 1 of 'unsigned char pgm_read_byte(const void*)'
  116 | static inline unsigned char pgm_read_byte(const void *addr) {
      |    
```

In der `pgmspace.h` steht
```C++
static inline unsigned char pgm_read_byte(const void *addr) {
    return *(const unsigned char *)(addr);
}
```

Hast du einen Tipp?

EDIT: Hat sich glaub ich erledigt.
Das Problem ist, dass der RP2040 gar kein PROGMEM kennt
https://forum.arduino.cc/t/rp2040-progmem-arrays/866697/5

**Comment by jp112sdl** [comment]
Also grundsätzlich ist da schon mal Leben drin.

<img width="340" alt="Bildschirmfoto 2022-03-02 um 21 17 36" src="https://user-images.githubusercontent.com/18190139/156442695-b3ebbea0-82db-44c7-aa14-c6a5595c7e0d.png">

Bei der seriellen Baudrate kann man einstellen was man will... es kommt immer mit 115200.
<img width="365" alt="Bildschirmfoto 2022-03-02 um 21 19 45" src="https://user-images.githubusercontent.com/18190139/156442971-8d38e5b1-b4b3-45b5-affd-d18b40566a07.png">
Aber ich glaub, damit kann man leben.

Next step: die Timer-Geschichte anschauen, da momentan die AlarmClock noch nicht läuft.
Aber eins nach dem anderen.

**Reply by psi-4ward** [reply]
Ja hart, damit wäre das Teil ja schon sniffer ready? 😳 Beeindruckend

**Comment by jp112sdl** [comment]
Der aktuelle Stand meiner Arbeitskopie liegt hier im Branch:
https://github.com/jp112sdl/AskSinPP/tree/dev_rp2040

Zu verwendende Board-Unterstützung:
https://github.com/earlephilhower/arduino-pico
Zusätzlich benötigte Timer-Lib:
https://github.com/khoih-prog/RPI_PICO_TimerInterrupt

Augenscheinlich ist es so für den Sniffer auf alle Fälle lauffähig.

Ich weiß nur nicht, ob die Änderung
https://github.com/jp112sdl/AskSinPP/blob/a18614897781c332b9b4a7559b8eaf0e244d4a58/AskSinPP.h#L91
```C++
      *dest = pgm_read_byte((uint16_t*)(adr + i));
```
Auswirkungen auf die anderen µC hat.

Sonst müsste man die Zeile noch mal extra in ein #ifdef packen.

**Reply by jp112sdl** [reply]
FreqTest bzw. EEPROM Schreiben/Lesen habe ich bisher noch nicht getestet!
Habe bei meinem Versuchsaufbau ein eByte Modul verwendet.

**Reply by codmpm** [reply]
Ich muss wohl mal RP2040 ordern. Klasse Arbeit!

**Reply by jp112sdl** [reply]
EEPROM Schreiben/Lesen klappt.

Das OLED Display am AnalyzerSniffer sollte auch laufen.

**Comment by pa-pa** [comment]
Ahm .... Guten Morgen ....

Das sieht ja Klasse aus. Die Änderung würde ich erstmal mit einem #ifdef
isolieren. Solche Cast können komische Nebenwirkungen haben, da ja hier
eine Zahlenwert von Typ Interger 16bit in einen Pointer umgewandelt wird.
Warum hast Du nicht nach (const void*) gecastet ?

Muss ich mir auch mal so ein Ding bei Gelegenheit mitbestellen :-)

Grüße
  Holger

Am Do., 3. März 2022 um 06:59 Uhr schrieb Jérôme ***@***.***>:

**Reply by jp112sdl** [reply]
Ja stimmt... weiß auch nicht, was mich da geritten hat. Zumal es die Fehlermeldung ja schon so angemeckert hat.
Hab es jetzt so gemacht und noch ein paar weitere Dinge repariert (Hardware Serial lesen und attachInterrupts wie beim ESP32)

**Comment by TomMajor** [comment]
in den E-Mails die diesem Topic hier vorausgingen (Christoph/Patrik wegen AskSin Analyzer Version mit 32u4) - kam da nicht irgendwie die Info dass 32u4 **end-of-life** sein soll ?
kann das bitte noch mal detailliert werden von dem die Info kam? Finde dazu nicht viel im Netz.

**Reply by jp112sdl** [reply]
Ich find dazu auch nix

**Comment by TomMajor** [comment]
Hmm, ich habe die mail nicht mehr, aber glaube mich dunkel zu erinnern dass es @codmpm war mit der 32u4 end-of-life info, vlt. kann sich Patrik dazu noch mal äußern?, Danke.

**Comment by codmpm** [comment]
Bin mir da gerade unsicher... ich meine es gelesen zu haben, weiß aber nicht mehr wo. Aber ich habe die letzten Wochen so viele Distributoren-Seiten gesehen, dass ich da mit Sicherheit auch was verwechselt haben kann. Vielleicht wars auch irgend eine Varianten oder in einem bestimmten Gehäuse?

Habe auch in unseren Mails nur gefunden, dass ich die Frage in den Raum gestellt habe: 
> Hab ich das eben richtig gelesen, dass der 32U4 EOL ist?

Da bei MicroChiop nichts von EOL steht, gehe ich also nicht davon aus.
https://www.microchip.com/en-us/product/ATmega32U4

**Comment by codmpm** [comment]
Achso, hab mal ein Board für den AskSinAnalyzer auf Basis des RP2040 designt: https://a360.co/35hDyhH

Meine RP2040 müssten morgen kommen, dann kann ich die Schaltung mal aufbauen und testen.

**Reply by jp112sdl** [reply]
Sieht schon schick aus. 
Den BOOT-Taster braucht es nicht unbedingt, aber wird den Kohl jetzt auch nicht fett machen. 
Man bekommt den RP2040 auch in den Bootloader-Modus, wenn man die Baudrate auf 1200 setzt (No-Button-Boot).
https://forums.raspberrypi.com/viewtopic.php?t=302144&start=50

**Reply by codmpm** [reply]
Ja, das kommt dann auch nicht mehr drauf an und ich denke mit Button ist "more convenient".

**Comment by TomMajor** [comment]
Danke für die Klarstellung bzgl. 32u4, Patrik. Dann ist das Thema vom Tisch. 😸 
p.s. nettes Board.

**Comment by codmpm** [comment]
Sooo, nach tatkräftiger Unterstützung von @jp112sdl habe ich das Ganze mit PlattformIO zum laufen bekommen.

Grundlage bildet der RP2040/Pico Core von [earlephilhower](https://github.com/earlephilhower/arduino-pico). Anscheinend gibt es für PlattformIO wohl noch keinen "offiziellen" Arduino RP2040 Core - korrigiert mich da gerne. 
Ich habe den Core von https://github.com/maxgerhardt/platform-raspberrypi genutzt. Leider muss man die ToolChain entsprechend seiner Plattform hinterlegen, also bei mir `darwin`. Alles Releases findet ihr hier: https://github.com/earlephilhower/pico-quick-toolchain/releases/ -  als einfach die URL bei `toolchain-pico@` anpassen.

I2C und SPI habe ich auf die defaults gehangen (dunklere Färbung im Pinout): https://www.raspberrypi.com/documentation/microcontrollers/raspberry-pi-pico.html
`GDO0` auf GP2, so wie bei @TomMajor's [AskSin Analyzer board](https://github.com/TomMajor/SmartHome/tree/master/PCB/AskSin-Analyzer-XS) - weiß nur nicht ob das so richtig ist.

`#include <EnableInterrupt.h>` muss in `AskSinSchniffer328P.ino` auskommentiert werden. Kann man sicher auch mit nem `ifdef` machen ;-)

Anscheinend klappt es auch nicht immer. Manchmal steht das Display und es kommt nichts per Serial, erst nach mehrmaligen Beustarts. Die LED blinkt immer freudig am Anfang.

Hier mal die Board-Definitions für die `plattformio.ini` entsprechend mit Jeromes fork: 
```
[env:rp2040]
monitor_speed = 115200
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
board_build.core = earlephilhower
board_build.filesystem_size = 1M
lib_ldf_mode = chain+
platform_packages = 
    maxgerhardt/framework-arduinopico@https://github.com/maxgerhardt/arduino-pico.git
    maxgerhardt/toolchain-pico@https://github.com/earlephilhower/pico-quick-toolchain/releases/download/1.3.3-a/x86_64-apple-darwin14.arm-none-eabi-ed6d983.220212.tar.gz
lib_deps = 
  RPI_PICO_TimerInterrupt
  AskSinPP=https://github.com/jp112sdl/AskSinPP/archive/refs/heads/dev_rp2040.zip


[env:rp2040-oled]
monitor_speed = 115200
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
board_build.core = earlephilhower
build_flags =
  -D USE_OLED
  -D LED_PIN=8
board_build.filesystem_size = 1M
lib_ldf_mode = chain+
platform_packages = 
    maxgerhardt/framework-arduinopico@https://github.com/maxgerhardt/arduino-pico.git
    maxgerhardt/toolchain-pico@https://github.com/earlephilhower/pico-quick-toolchain/releases/download/1.3.3-a/x86_64-apple-darwin14.arm-none-eabi-ed6d983.220212.tar.gz
lib_deps = 
  RPI_PICO_TimerInterrupt
  AskSinPP=https://github.com/jp112sdl/AskSinPP/archive/refs/heads/dev_rp2040.zip
  Adafruit GFX Library
  Adafruit SSD1306
```


--- 
![asksin-cc1101-rp2040-sniffer-prototyp](https://user-images.githubusercontent.com/9745239/158634172-5180ac26-6c72-4ef5-b26c-23535bb24131.jpeg)

Im Grund funktioniert es auch schon super, allerdings sieht mir das output am Serial noch etwas komisch aus. Ach ja, muss `115200` sein, wie Jerome schon schrieb.

```
Packet too big: 86
Packet too big: 139
Packet too big: 127
Packet too big: 116
Packet too big: 161
Packet too big: 139
Packet too big: 127
Packet too big: 116
Packet too big: 161
CRC Failed
Packet too big: 86
Packet too big: 31
Packet too big: 147
```

@psi-4ward's Analyzer zeigt auch lustig Daten an, allerdings angeblich HMIP (was hier im Büro nicht verbaut ist) und gar nichts, wenn ich eine alte HM-Steckdose per Homegear schalte.

**Reply by meier81** [reply]
> Sooo, nach tatkräftiger Unterstützung von @jp112sdl habe ich das Ganze mit PlattformIO zum laufen bekommen.

Hi codmpm bzw. euch allen,

dachte ich nehme mich dem Projekt auch mal an und probiere ein bisschen, will aber irgendwie nicht zur Zeit. Hab die Definition der platform.ini von codmpm genommen, er sollte sich dadurch ja die benötigen packages runterladen bzw. installieren. Hab hier bei mir Windows 10 64bit und Visual Studio Code, bin mir da nicht sicher welche Toolchain hier richtig ist, hab die x86_64-w64-mingw32.arm-none-eabi-ed6d983.220212.zip benutzt, scheint aber wohl nicht die richtige zu sein.

Könnt ihr mich hier mal abholen bezüglich der richtigen Toolchain und evtl. Tipps und Tricks beim einrichten?

Vielen Dank jedenfalls schonmal.

Gruß Markus

**Reply by codmpm** [reply]
Ich geb zu da habe ich auch nicht durchgeblickt. Habe quasi "mit Glück" die richtige für OSX gegriffen. Hätte jetzt gesagt, dass [x86_64-w64-mingw32.arm-none-eabi-ed6d983.220212.zip](https://github.com/earlephilhower/pico-quick-toolchain/releases/download/1.3.3-a/x86_64-w64-mingw32.arm-none-eabi-ed6d983.220212.zip) die richtige ist.

Plattform.io hast du aber natürlich vorher installiert in VSCode, oder?

Hab gerade leider kein Windows zum gegentesten hier weil M1 Mac.
Ansonsten nimm doch die Arduino-IDE. Da sollte es doch drin klappen, oder?

**Reply by meier81** [reply]
Denke ich schreib dir mal meine "Basis":

Wie gesagt, Windows 10 64bit, Visual Studio Code aktuell mit installiertem Plattform.IO.
Hab mir zuerst ein neues Verzeichnis angelegt und darin deine plattformio.ini von diesem Post angelegt und die Toolchain angepasst sieht also jetzt so aus:

`[env:rp2040]
monitor_speed = 115200
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
board_build.core = earlephilhower
board_build.filesystem_size = 1M
lib_ldf_mode = chain+
platform_packages = 
    maxgerhardt/framework-arduinopico@https://github.com/maxgerhardt/arduino-pico.git
    maxgerhardt/toolchain-pico@https://github.com/earlephilhower/pico-quick-toolchain/releases/download/1.3.3-a/x86_64-w64-mingw32.arm-none-eabi-ed6d983.220212.zip
lib_deps = 
  RPI_PICO_TimerInterrupt
  AskSinPP=https://github.com/jp112sdl/AskSinPP/archive/refs/heads/dev_rp2040.zip


[env:rp2040-oled]
monitor_speed = 115200
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
board_build.core = earlephilhower
build_flags =
  -D USE_OLED
  -D LED_PIN=8
board_build.filesystem_size = 1M
lib_ldf_mode = chain+
platform_packages = 
    maxgerhardt/framework-arduinopico@https://github.com/maxgerhardt/arduino-pico.git
    maxgerhardt/toolchain-pico@https://github.com/earlephilhower/pico-quick-toolchain/releases/download/1.3.3-a/x86_64-w64-mingw32.arm-none-eabi-ed6d983.220212.zip
lib_deps = 
  RPI_PICO_TimerInterrupt
  AskSinPP=https://github.com/jp112sdl/AskSinPP/archive/refs/heads/dev_rp2040.zip
  Adafruit GFX Library
  Adafruit SSD1306`

Nun habe ich die .ino Datei erstellt mit dem Inhalt einen Post weiter unten von Jérôme.

Dann in Platform.IO das Projekt geöffnet (muss ja eine kurze Zeit warten, er lädt ja nun die Dateien nach) und gehe dann auf PlatformIO:Build. Bekomme dann folgenden Fehler im Terminal:

`> Executing task: C:\Users\Familie\.platformio\penv\Scripts\platformio.exe run --environment rp2040 <

Processing rp2040 (platform: https://github.com/maxgerhardt/platform-raspberrypi.git; board: pico)
-------------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/raspberrypi/pico.html
PLATFORM: Raspberry Pi RP2040 (1.6.0+sha.d20e2ed) > Raspberry Pi Pico
HARDWARE: RP2040 133MHz, 264KB RAM, 2MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, jlink, picoprobe, raspberrypi-swd)
PACKAGES:
 - framework-arduinopico 1.10914.0+sha.beec911
 - tool-rp2040tools 1.0.2
 - toolchain-gccarmnoneeabi 1.90201.191206 (9.2.1)
 - toolchain-pico 5.100300.220212 (10.3.0)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain+, Compatibility ~ soft
Found 1 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <RPI_PICO_TimerInterrupt> 1.2.0
Error: Nothing to build. Please put your source code files to 'C:\Users\Familie\Documents\PlatformIO\Projects\RP2040\src' folder
================================================= [FAILED] Took 1.25 seconds =================================================

Environment    Status    Duration
-------------  --------  ------------
rp2040         FAILED    00:00:01.246
============================================ 1 failed, 0 succeeded in 00:00:01.246 ============================================Der Terminalprozess "C:\Users\Familie\.platformio\penv\Scripts\platformio.exe 'run', '--environment', 'rp2040'" wurde mit folgendem Exitcode beendet: 1.`

Mach ich hier was falsch beim bauen?

Gruß Markus

**Comment by jp112sdl** [comment]
> #include <EnableInterrupt.h> muss in AskSinSchniffer328P.ino auskommentiert werden. Kann man sicher auch mit nem ifdef machen ;-)

Ich hatte den XS-Sketch für den Pi Pico schon angepasst und Psi gesendet. Oder doch nicht?
```
//- -----------------------------------------------------------------------------------------------------------------------
// AskSin++
// 2016-10-31 papa Creative Commons - http://creativecommons.org/licenses/by-nc-sa/3.0/de/
// 2019-05-24 jp112sdl Creative Commons - http://creativecommons.org/licenses/by-nc-sa/3.0/de/
//- -----------------------------------------------------------------------------------------------------------------------

// Enable th 0.91" 128x32 OLED display
// For use with RP2040 comment these in lines in Adafruit_SSD1306.cpp:
//  #define pgm_read_byte(addr)                                                    \
//    (*(const unsigned char *)(addr)) ///< PROGMEM workaround for non-AVR
// #define USE_OLED


#include <SPI.h>
#ifdef ARDUINO_ARCH_AVR
#define EI_NOTEXTERNAL
#include <EnableInterrupt.h>
#endif
#include <AskSinPP.h>
#include <Device.h>
#include <Register.h>

#ifdef USE_OLED
#include "Ssd1306.h"
#include <Adafruit_I2CDevice.h>
#endif

#ifdef ARDUINO_ARCH_AVR
  #ifndef LED_PIN
    #define LED_PIN  4
  #endif
#define CS_PIN   10
#define GDO0_PIN 2
#else
#define LED_PIN  D25 // Builtin LED on Raspbbery Pi Pico
#define CS_PIN   SS
#define GDO0_PIN D20
#endif

#ifdef SSD1306_H_
#define PEAK_THRESHOLD 5 // count the RSSI Polls for the peak threshold
#endif

#define RSSI_POLL_INTERVAL 200 // poll every x milliseconds
#define RSSI_SERIAL_FACTOR 5   // sends all RSSI_POLL_INTERVAL * RSSI_SERIAL_FACTOR the lowest RSSI to Serial

// all library classes are placed in the namespace 'as'
using namespace as;

// define all device properties
const struct DeviceInfo PROGMEM devinfo = {
  {0xFF, 0xFF, 0xFF},     // Device ID
  "..........",           // Device Serial
  {0x00, 0x00},           // Device Model
  0x10,                   // Firmware Version
  as::DeviceType::Remote, // Device Type
  {0x00, 0x00}            // Info Bytes
};

typedef AskSin<StatusLed<LED_PIN>, NoBattery, Radio<LibSPI<CS_PIN>, GDO0_PIN>> HalType;

class SnifferDevice : public Device<HalType, DefList0>, Alarm {
  DefList0 l0;
  
  private: 
    uint8_t cnt;
    uint8_t rssi_peak;
    uint8_t rssi_act; 
  public:
    typedef Device<HalType, DefList0> BaseDevice;
#ifdef SSD1306_H_
    DisplayType<PEAK_THRESHOLD, 0x3C> display;
#endif
    SnifferDevice (const DeviceInfo& i, uint16_t addr) : BaseDevice(i, addr, l0, 0), Alarm(0), l0(addr)  {}
    virtual ~SnifferDevice () {}

    virtual void trigger (__attribute__ ((unused)) AlarmClock& clock) {
      set(millis2ticks(RSSI_POLL_INTERVAL));
      clock.add(*this);
      this->radio().pollRSSI();
      cnt++;
      rssi_act = this->radio().rssi();
      if (rssi_act < rssi_peak) {
        rssi_peak = rssi_act;
      }
      if (cnt >= RSSI_SERIAL_FACTOR) {
        DPRINT(":");DHEX(rssi_peak);DPRINTLN(";"); 
        cnt = 0;
        rssi_act = 0;
        rssi_peak = 199;
      }
#ifdef SSD1306_H_
      display.printFull(this->radio().rssi());
#endif
    }

    virtual bool process(Message& msg) {
      DPRINT(F(":"));
      DHEX(radio().rssi());
      DHEX(msg.length());
      DHEX(msg.count());
      DHEX(msg.flags());
      DHEX(msg.type());
      msg.from().dump();
      msg.to().dump();
      for (uint8_t l = 0; l < msg.length() - 9; l++) DHEX(msg.buffer()[l + 9]);
      DPRINTLN(";");
      this->led().ledOn(millis2ticks(100));
      return true;
    }

    bool init (HalType& hal) {
      HMID id;
      this->setHal(hal);
      this->getDeviceID(id);
      hal.init(id);
      hal.config(this->getConfigArea());
      sysclock.add(*this);
      return false;
    }
};

HalType hal;
SnifferDevice sdev(devinfo, 0x20);

void setup () {
#ifdef ARDUINO_ARCH_RP2040
  delay(2500);
#endif
  DINIT(57600, ASKSIN_PLUS_PLUS_IDENTIFIER);
#ifdef SSD1306_H_
  sdev.display.init();
#endif
  sdev.init(hal);
}

void loop() {
  hal.runready();
  sdev.pollRadio();
}
```

**Reply by codmpm** [reply]
Sorry, hatte den Master genommen. Probiere das morgen aus 👍

**Comment by psi-4ward** [comment]
Ich hab grad aus privaten Gründen keine Zeit für Hobby

**Comment by codmpm** [comment]
Mit @jp112sdl's code von oben läufts komplett. Offensichtlich war der in meiner Version fehlende `GDO0` das Problem.
Dann mach ich mal den Prototyp der Platine fertig 😬👍🏻

Hier sieht man schön wie meine alte HM-LC-Sw1-Pl-2 schaltet.
![Bildschirmfoto 2022-03-17 um 10 55 24](https://user-images.githubusercontent.com/9745239/158784510-922ab866-1823-4df6-98b2-26b3f16ad704.png)

**Comment by codmpm** [comment]
Mag nochmal jemand drüber gucken, bevor ich die Prototypen bestelle? 

![cc1101_sniffer_with_display-codm](https://user-images.githubusercontent.com/9745239/158990976-41d8a287-c10d-46c3-8239-28cdd1ebc6af.png)


3D-Ansicht: https://a360.co/37AgyLn
Aisler: https://aisler.net/p/UKZUBLAQ
Schaltplan: [cc1101-sniffer.sch.pdf](https://github.com/pa-pa/AskSinPP/files/8303577/cc1101-sniffer.sch.pdf)

**Comment by codmpm** [comment]
Hab mal ein Gehäuse gebaut: https://a360.co/3tA1E0u
Über den Modell-Browser kann man einzelne Teile ausblenden.

![codm-cc1101-sniffer-case](https://user-images.githubusercontent.com/9745239/159693164-f8c5ea06-fe82-46d4-bcb0-c4e4d501dab0.png)

**Comment by codmpm** [comment]
![cc1101-rp2040-sniffer-naked-0 1](https://user-images.githubusercontent.com/9745239/160862559-fd152450-cce6-4462-808c-329e0d5f1118.jpg)
![cc1101-rp2040-sniffer-display-0 1](https://user-images.githubusercontent.com/9745239/160862569-85e1d92c-1b24-40df-8144-e59880844380.jpg)

Läuft!!! 🥳
Ich denke ich kann ein paar 100nF runter nehmen und das Display müsste noch geflippt werden. 😉

Warte aktuell auf die ebyte E07-900M10S weil es die E07-868MS10 (rotes Modul) gerade nur im 500er Pack gibt. Sollten diese Woche kommen, da ich sowieso eine größere Bestellung bei ebyte wegen Antennen hatte.
@TomMajor scheint damit ja schon mal Kontakt gehabt zu haben: https://forum.fhem.de/index.php?topic=126449.0

Werde die dann testen und ne V0.2, am besten mit dem 900MS10 machen.

**Reply by TomMajor** [reply]
yep, hatte es auch hier gepostet
https://homematic-forum.de/forum/viewtopic.php?f=76&t=71119&p=709379#p709379
Die Info kam anscheinend direkt vom Hersteller das E07-868MS10 End of Life sein soll.
ich denke das neue E07-900M10S wird gut funktionieren, sieht zumindest hochwertig aus.

**Reply by codmpm** [reply]
Die Info habe ich auch von denen direkt. 500 wollte er mir verkaufen, weil sie die dann für mich bauen würden.
Ich gucke wie die 900MS10S sich schlagen, mir sieht das qualitativ auch in Ordnung aus. Hatte die E18 (zigbee) von denen die eine ähnliche Bauform haben.

**Reply by TomMajor** [reply]
ja mach mal, bin gespannt wie die 900MS10S dann aus dem Test rauskommen.
FreqTest wäre auch interessant, und zwar die dB Werte, im Vergleich mit einem "normalen" CC1101 Modul - am gleichen Standort und mit gleicher Antenne.
Nettes Board Design, btw 😸

**Comment by codmpm** [comment]
![ebyte-e07-900m10s](https://user-images.githubusercontent.com/9745239/161374464-7b93536c-6920-40f3-a123-3d0fcb94f776.png)

Die E07-900M10S sind gestern schon gekommen. Hab mal ne Eagle-Lib mit 3D-Modell gemacht. Kann man doch in KiCad importieren, oder? 
Melde mich sowie ich zum Testen der 900M10S gekommen bin.

LBR: https://github.com/codm/eagle-libraries#e07-900m10s=
3D-Modell: https://grabcad.com/library/e07-900m10s-1

**Reply by codmpm** [reply]
![IMG_9677 Kopie](https://user-images.githubusercontent.com/9745239/161379384-9d12f0d7-6846-452c-9b4d-f4b4d9ff5ba2.jpg)

Läuft! Quasi nur umgesteckt.

Wie kann ich jetzt den Frequenzvergleich machen?

**Reply by codmpm** [reply]
![codm-ccc1101-sniffer-0 2](https://user-images.githubusercontent.com/9745239/161381924-4ded290c-b503-4da1-9d38-32d52dc473e8.png)

**Reply by TomMajor** [reply]
ich meinte mal die dBm Werte vergleichen wenn beide Module denselben Standort und damit Entfernung zur Zentrale haben, allerdings müssten sie auch die gleiche Antenne haben.
Das wird aber schwierig wie ich sehe wegen der unterschiedl. Antennenkontakierungen..

**Comment by meier81** [comment]
Hallo euch allen,

so, ich hab das jetzt auch mit dem Raspberry Pico zu laufen bekommen, zwar mit der Arduino IDE aber gut, läuft ja.

Hab bei mir den HM-LC-Bl1-FM aufgebaut auf dem Steckbrett, lässt sich auch soweit anlernen und bedienen über die CCU, habe allerdings das Problem das ich eine Einstellung des Gerätes vorgenommen habe, diese hat er auch korrekt gesendet aber anscheinend keine Antwort bekommen, deshalb steht die Meldung "Konfigurationsdaten stehen zur Übertragung an" noch an, er bekommt da anscheinend keine Rückmeldung das er sie korrekt gesendet hat. Funktioniert auch nach mehrmaligem probieren nicht.

Zudem kommt die Meldung "Gerätekommunikation gestört" kurz nachdem ich eine Fahrt über die CCU bzw. über die Taster vorgenommen habe, wobei die Kommunikation trotzdem einwandfrei funktioniert.

Hier mal mein verwendeter Sketch:
```
//- -----------------------------------------------------------------------------------------------------------------------
// AskSin++
// 2017-12-14 papa Creative Commons - http://creativecommons.org/licenses/by-nc-sa/3.0/de/
// ci-test=yes board=rp2040 aes=no
//- -----------------------------------------------------------------------------------------------------------------------

// define this to read the device id, serial and device type from bootloader section
// #define USE_OTA_BOOTLOADER

#include <SPI.h>
#include <AskSinPP.h>
#include <Blind.h>

#define LED_PIN D25 // Builtin LED on Raspbbery Pi Pico

#define CONFIG_BUTTON_PIN D6

#define CC1101_CS_PIN SS
#define CC1101_GDO0_PIN D20

#define UP_RELAY_PIN D8
#define DOWN_RELAY_PIN D9

#define UP_BUTTON_PIN D10
#define DOWN_BUTTON_PIN D11

// number of available peers per channel
#define PEERS_PER_CHANNEL 6

// all library classes are placed in the namespace 'as'
using namespace as;

// define all device properties
const struct DeviceInfo PROGMEM devinfo = {
    {0x59,0x32,0xaf},              // Device ID
    "papa5932af",                  // Device Serial
    {0x00,0x05},                   // Device Model
    0x24,                          // Firmware Version
    as::DeviceType::BlindActuator, // Device Type
    {0x01,0x00}                    // Info Bytes
};

/**
 * Configure the used hardware
 */
typedef LibSPI<CC1101_CS_PIN> RadioSPI;
typedef AskSin<StatusLed<LED_PIN>,NoBattery,Radio<RadioSPI,CC1101_GDO0_PIN> > Hal;

DEFREGISTER(BlindReg0,MASTERID_REGS,DREG_INTKEY,DREG_CONFBUTTONTIME,DREG_LOCALRESETDISABLE)

class BlindList0 : public RegList0<BlindReg0> {
public:
  BlindList0 (uint16_t addr) : RegList0<BlindReg0>(addr) {}
  void defaults () {
    clear();
    // intKeyVisible(false);
    confButtonTime(0xff);
    // localResetDisable(false);
  }
};

class BlChannel : public ActorChannel<Hal,BlindList1,BlindList3,PEERS_PER_CHANNEL,BlindList0,BlindStateMachine> {
private:
  uint8_t up_relay_pin;
  uint8_t down_relay_pin;
public:
  typedef ActorChannel<Hal,BlindList1,BlindList3,PEERS_PER_CHANNEL,BlindList0,BlindStateMachine> BaseChannel;

  BlChannel () : up_relay_pin(0), down_relay_pin(0) {}
  virtual ~BlChannel () {}

  virtual void switchState(uint8_t oldstate,uint8_t newstate, uint32_t stateDelay) {
    BaseChannel::switchState(oldstate, newstate, stateDelay);
    if( newstate == AS_CM_JT_RAMPON && stateDelay > 0 ) {
      motorUp();
    }
    else if( newstate == AS_CM_JT_RAMPOFF && stateDelay > 0 ) {
      motorDown();
    }
    else {
      motorStop();
    }
  }

  void motorUp () {
    digitalWrite(down_relay_pin,LOW);
    digitalWrite(up_relay_pin,HIGH);
  }

  void motorDown () {
    digitalWrite(up_relay_pin,LOW);
    digitalWrite(down_relay_pin,HIGH);
  }

  void motorStop () {
    digitalWrite(up_relay_pin,LOW);
    digitalWrite(down_relay_pin,LOW);
  }

  void init (uint8_t up, uint8_t dp) {
    up_relay_pin = up;
    down_relay_pin = dp;
    pinMode(up_relay_pin,OUTPUT);
    pinMode(down_relay_pin,OUTPUT);
    motorStop();
    BaseChannel::init();
  }
};

// setup the device with channel type and number of channels
typedef MultiChannelDevice<Hal,BlChannel,1,BlindList0> BlindType;

Hal hal;
BlindType sdev(devinfo,0x20);
ConfigButton<BlindType> cfgBtn(sdev);
InternalButton<BlindType> btnup(sdev,1);
InternalButton<BlindType> btndown(sdev,2);

void initPeerings (bool first) {
  // create internal peerings - CCU2 needs this
  if( first == true ) {
    sdev.channel(1).peer(btnup.peer(),btndown.peer());
  }
}

void setup () {
  DINIT(57600,ASKSIN_PLUS_PLUS_IDENTIFIER);
  bool first = sdev.init(hal);
  sdev.channel(1).init(UP_RELAY_PIN, DOWN_RELAY_PIN);

  buttonISR(cfgBtn,CONFIG_BUTTON_PIN);
  buttonISR(btnup,UP_BUTTON_PIN);
  buttonISR(btndown,DOWN_BUTTON_PIN);

  initPeerings(first);
  sdev.initDone();
}

void loop() {
  hal.runready();
  sdev.pollRadio();
}
``` 

Hier mal ein Bild von meinen Arduino IDE Einstellungen:
![Arduino IDE](https://user-images.githubusercontent.com/40598096/168671345-17f9abaf-7e89-4d12-8eb8-381d597c7dd7.JPG)

Und anbei noch ein Auszug des seriellen Logs:
[Log.txt](https://github.com/pa-pa/AskSinPP/files/8703081/Log.txt)

Gruß Markus

**Reply by meier81** [reply]
Kurzes Update:

Hab eben mal den Pico neu gestartet, Meldungen sind weg, läuft erst mal einwandfrei bis jetzt. Werde morgen mal weiter testen und berichten!
